### DISCUSSION TITLE
Gulp src not matching nested directories

### DISCUSSION THREAD

**Discussion by webstackdev** [gulpjs/gulp#2452] [Help]
*I posted this on [Stack Overflow](https://stackoverflow.com/questions/62289044/gulp-src-not-matching-nested-directories) also, please let me know if I should only post here or only on SO for this kind of question in the future. I can't tell if this is a bug or (more likely) my lack of a clue.*

    gulp: v4.0.0
    gulp-cli: v2.1.0
    gulp-debug: v4.0.0
    glob: v7.1.6
    node: v10.16.0
    OS: Ubuntu 18.04

I have these two functions in the same gulpfile:

    function debugGulpSrc() {
      ['/var/abc/scss/**/*.scss'].forEach( (globPattern) => {
      return src( globPattern )
        .pipe( debug({minimal: true}) )
      })
    }

    function debugNodeGlob() {
      ['/var/abc/scss/**/*.scss'].forEach( (globPattern) => {
        glob( globPattern, {}, function (err, files) {
          if (err) {
            console.log('Error: ' + err);
            return;
          }

          files.forEach( (file) => {
            console.log(file);
          });
        })
      })
    }

The single entry in the glob array is intentional. I've been testing with just one but once it works I will add more glob patterns from the project root. I'm using the `forEach` because Gulp's `src()` function errors with no warning (even with `strict`on and `silent` off) if there are no matches to the pattern. It's a large project and I have some folders that will not match a pattern until later on.

The file structure I'm trying to glob looks like this:

    /var/abc/scss/style.scss
    /var/abc/scss/folder1/style.scss
    /var/abc/scss/folder1/index.html
    /var/abc/scss/folder1/nested_dir/index.scss
    /var/abc/scss/folder2/style.scss
    /var/abc/scss/folder2/index.html
    ... many other folders and files

My `debugNodeGlob()` function (using `glob`) returns all of the files I expect. However `debugGulpSrc()` returns the files in the root directory and the root of `folder1`, but nothing else. It does not get the files in nested directories. The Gulp src documentation says that it should with the double-star (`**`) glob pattern, and `glob` on its own does so normally.

I `diff`'d the `debug` output set as an option in both approaches, and everything was identical except the output.

The only difference I can see (except for many files not returned by Gulp src) is that the files returned by `glob` have absolute paths, and those returned by `src` have relative paths to the project root:

    /var/abc/scss/folder1/style.scss

vs.

    ../abc/scss/folder1/style.scss

I am running the script file from here from the command line:

    /var/abc/scripts/gulpfile.js

I use the same glob patterns as input to `gulp-watch`, and it works fine and handles nested subdirectories correctly.

I have a feeling that Gulp `src` is doing something where it expands my double-star wildcard to every directory actually on the block device, finds one without a match, and dies silently, but I have no idea how to fix that or get the functionality that I need. It's not reasonable to add a new glob pattern to the array every time I add a subdirectory. The double-star glob pattern should handle that (including subfolders that don't have matching files), just like Isaac's `glob` library does (the documentation says `src` is a wrapper around `glob`).

I haven't grasped why Gulp doesn't offer an option to continue with the next glob pattern in an array of globs after one doesn't match, and instead chooses to die without emitting an event I could do something with. If it does emit an event, it's not in the documentation and isn't caught with .on(error, fn). If anyone has insight into that design decision I'd appreciate it.

**Comment by phated** [comment]
Hey @calligraphic, it's fine to post questions in the gulp repo, but we have Discussions enabled now so I'm going to move this over there.

As for your question, gulp.src() is a stream and you aren't draining it so I wouldn't expect you to be getting everything.
